### PR TITLE
Add CVE and CWE roles

### PR DESCRIPTION
**Pull request by hugovk**
Subject: <short purpose of this pull request>

### Feature or Bugfix

- Feature

### Purpose
- Add a `cve` role to turn ``` :cve:`2020-10735` ``` into https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-10735
- Add a `cwe` role to turn ``` :cwe:`787` ``` into https://cwe.mitre.org/data/definitions/787.html

### Detail
- Add `cve` and `cwe` roles, similar to existing `pep` and `rfc` roles, to link to the Common Vulnerabilities and Exposures and Common Weakness Enumeration websites.

### Relates
- https://github.com/sphinx-doc/sphinx/pull/11780

### CODE REVIEW

File: sphinx/roles.py [hunk 1]
```diff
@@ -172,6 +172,78 @@ def process_link(self, env: BuildEnvironment, refnode: Element, has_explicit_tit
...
+        target = nodes.target("", "", ids=[target_id])
+        self.inliner.document.note_explicit_target(target)
+
+        try:
+            refuri = self.build_uri()
+            reference = nodes.reference("", "", internal=False, refuri=refuri, classes=["cve"])
+            if self.has_explicit_title:
+                reference += nodes.strong(self.title, self.title)
+            else:
+                title = "CVE " + self.title
```
Comment (picnixz) [inline sphinx/roles.py]:
```suggestion
                title = f"CVE {self.title}"
```

File: sphinx/roles.py [hunk 2]
```diff
@@ -172,6 +172,78 @@ def process_link(self, env: BuildEnvironment, refnode: Element, has_explicit_tit
...
+
+        return [index, target, reference], []
+
+    def build_uri(self) -> str:
+        base_url = self.inliner.document.settings.cve_base_url
+        ret = self.target.split("#", 1)
+        if len(ret) == 2:
+            return f"{base_url}{ret[0]}#{ret[1]}"
+        else:
+            return f"{base_url}{ret[0]}"
```
Comment (picnixz) [inline sphinx/roles.py]:
```suggestion
        if len(ret) == 2:
            return f"{base_url}{ret[0]}#{ret[1]}"
        return f"{base_url}{ret[0]}"
```

File: sphinx/roles.py [hunk 3]
```diff
@@ -172,6 +172,78 @@ def process_link(self, env: BuildEnvironment, refnode: Element, has_explicit_tit
...
+
+
+class CWE(ReferenceRole):
```
Comment (picnixz) [inline sphinx/roles.py:243]: Same comments as for CVE I think (you probably CC right?)

File: sphinx/roles.py [hunk 4]
```diff
@@ -172,6 +172,78 @@ def process_link(self, env: BuildEnvironment, refnode: Element, has_explicit_tit
...
+        target = nodes.target("", "", ids=[target_id])
+        self.inliner.document.note_explicit_target(target)
+
+        try:
+            refuri = self.build_uri()
+            reference = nodes.reference("", "", internal=False, refuri=refuri, classes=["cwe"])
+            if self.has_explicit_title:
+                reference += nodes.strong(self.title, self.title)
+            else:
+                title = "CWE " + self.title
```
Comment (hugovk) [inline sphinx/roles.py]:
```suggestion
                title = f"CWE {self.title}"
```

File: sphinx/roles.py [hunk 5]
```diff
@@ -172,6 +172,78 @@ def process_link(self, env: BuildEnvironment, refnode: Element, has_explicit_tit
...
+
+        return [index, target, reference], []
+
+    def build_uri(self) -> str:
+        base_url = self.inliner.document.settings.cwe_base_url
+        ret = self.target.split("#", 1)
+        if len(ret) == 2:
+            return f"{base_url}{int(ret[0])}.html#{ret[1]}"
+        else:
+            return f"{base_url}{int(ret[0])}.html"
```
Comment (hugovk) [inline sphinx/roles.py]:
```suggestion
        if len(ret) == 2:
            return f"{base_url}{int(ret[0])}.html#{ret[1]}"
        return f"{base_url}{int(ret[0])}.html"
```

**Comment by AA-Turner** [comment]
No objection in principle, though this feels like a feature is missing to make this easier -- these roles are just extlinks, in bold, with customised classes ('cve'/'cwe') and index entries.

A

**Comment by picnixz** [comment]
We could refactor this by adding a class reponsible for this kind of things maybe?

**Comment by AA-Turner** [comment]
cc @picnixz was anything outstanding on this?

**Review by AA-Turner — APPROVED** [review]

**Comment by picnixz** [comment]
Ok, so something happened in between. I'll need to check the markup though I don't think I've changed anything on that part.

**Comment by AA-Turner** [comment]
> Ok, so something happened in between. I'll need to check the markup though I don't think I've changed anything on that part.
> 
> So it appears that only the CWE role fails. Hum.

My fault, `v` and `w` are too similar...

a

**Comment by picnixz** [comment]
Oh. Yes the URL :') (dealing with CVE/CWE is the worst, there is only one letter differencing the two variables...)

File: doc/usage/restructuredtext/roles.rst [hunk 6]
```diff
@@ -249,6 +249,34 @@ There is also an :rst:role:`index` role to generate index entries.
 
 The following roles generate external links:
 
+.. rst:role:: cve
+
+   A reference to a `Common Vulnerabilities and Exposures`_ record.
+   This generates appropriate index entries.
+   The text "CVE *number*\ " is generated;
+   with a link to an online copy of the specified CVE.
+   You can link to a specific section by using ``:cve:`number#anchor```.
```
Comment (picnixz) [inline doc/usage/restructuredtext/roles.rst:258]: Don't we need something like:

```rst
``:cve:\`number#anchor\```
```

Or is Sphinx very smart (I don't remember)?

Comment (AA-Turner) [reply doc/usage/restructuredtext/roles.rst:258]: This works: https://sphinx--11781.org.readthedocs.build/en/11781/usage/restructuredtext/roles.html#role-cve

https://sphinx--11781.org.readthedocs.build/en/11781/usage/restructuredtext/roles.html#role-cwe

Comment (picnixz) [reply doc/usage/restructuredtext/roles.rst:258]: Sphinx is smarter than GitHub for that! (too many times when I needed to use HTML hacks on GH for this kind of things...)

**Review by picnixz — APPROVED** [review]

**Pull request merged**